### PR TITLE
http2: add optional connection ping/keepalive

### PR DESCRIPTION
**Pull request by ggreenway**
Signed-off-by: Greg Greenway <ggreenway@apple.com>

Commit Message:
Additional Description:
Risk Level: low (disabled by default)
Testing: Added new tests
Docs Changes: documented
Release Notes: added
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by ggreenway** [comment]
Should this be noted in the XDS docs somewhere as "we recommend setting this"? I didn't see a really good place to add it.

Any idea how to make the upstream integration test work? I expected what I wrote to work, but the test waits forever without the connection getting closed.

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:358]: Why must this be less? Can't it work similar to health checking where the interval and the timeout are disjoint? It seems like that might be what we want?

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:350]: Should we have any built-in jitter here, either just implicit or defined as a param?

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:349]: Thoughts on making this a dedicated message, with the sub-fields all required?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Nice, thanks for working on this. Just starting with a few API comments before diving into the code.

/wait

File: api/envoy/config/core/v3/protocol.proto
Comment (ggreenway) [reply api/envoy/config/core/v3/protocol.proto:358]: I did this for ease of implementation (so there can't be two outstanding pings at the same time). I figured it would easy to relax this at some point in the future.

File: api/envoy/config/core/v3/protocol.proto
Comment (ggreenway) [reply api/envoy/config/core/v3/protocol.proto:350]: Good point; we can probably borrow some of the stuff from health checks.

File: api/envoy/config/core/v3/protocol.proto
Comment (ggreenway) [reply api/envoy/config/core/v3/protocol.proto:349]: I don't care either way

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [reply api/envoy/config/core/v3/protocol.proto:358]: I'm not sure how you can guarantee that no matter what you do, due to event loop delay, etc. It seems like it would be better if it worked like health checking where there is an outstanding ping with a timeout, and then an interval.

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [reply api/envoy/config/core/v3/protocol.proto:350]: +1

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [reply api/envoy/config/core/v3/protocol.proto:349]: I think I would recommend that if you don't mind either way.

File: api/envoy/config/core/v3/protocol.proto
Comment (ggreenway) [reply api/envoy/config/core/v3/protocol.proto:358]: Yeah, I agree that doing it like the health checker will be better (and more consistent). I'll go that way.

File: source/common/http/http2/codec_impl_legacy.cc
Comment (yanavlasov) [inline source/common/http/http2/codec_impl_legacy.cc:580]: Looks like processing of the PING ACK is missing in the legacy codec.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

**Comment by yanavlasov** [comment]
It would also be good to add codec_impl_test.cc tests for the PINGs. There is a bit of an overlap with the integration tests, but having these lower level tests would be good too.

File: api/envoy/config/core/v3/protocol.proto
Comment (htuch) [reply api/envoy/config/core/v3/protocol.proto:349]: Yeah, +1 here.

File: source/common/http/http2/codec_impl.cc
Comment (htuch) [inline source/common/http/http2/codec_impl.cc:600]: How confident are you that this is always true?

Comment (htuch) [reply source/common/http/http2/codec_impl.cc:600]: I see, this is only NOMEM. In that case, I think a `RELEASE_ASSERT` is warranted.

Comment (ggreenway) [reply source/common/http/http2/codec_impl.cc:602]: I copied the pattern from elsewhere in the file; it's pervasive. I personally have no confidence that it's true.

**Comment by ggreenway** [comment]
@mattklein123 I added jitter, but it blew up the size of the PR, due to needing to pass `RandomGenerator` to the codec now. But I think it's worth it; jitter is rather important.

**Comment by mattklein123** [comment]
Can you merge main and check CI and I can take a look?

/wait

**Comment by ggreenway** [comment]
> Can you merge main and check CI and I can take a look?
> 
> /wait

I'm still trying to figure out a unit test (but I keep getting distracted by other things).

**Comment by ggreenway** [comment]
> It would also be good to add codec_impl_test.cc tests for the PINGs. There is a bit of an overlap with the integration tests, but having these lower level tests would be good too.

@yanavlasov I tried, but I can't figure out what/how the tests work. I think I can test the jitter, and that the initial timer is set, and what happens when the timeout timer fires, but I don't have any idea how to verify that a PING was either sent, or how to send/not-send a response PING.

**Comment by mattklein123** [comment]
> @yanavlasov I tried, but I can't figure out what/how the tests work. I think I can test the jitter, and that the initial timer is set, and what happens when the timeout timer fires, but I don't have any idea how to verify that a PING was either sent, or how to send/not-send a response PING.

The codec tests use a fake codec client for the other half of the connection. It should respond to pings just like a real connection does. Thus, if you don't hold any data, it should respond, and you for example held the the response data, it should not send any ping response. Whether this is worth it or not I will defer to both of you.

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:194]: s/interval_ms/interval here and below

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:197]: Can you clarify that default/zero means no jitter in the comment?

File: api/envoy/config/core/v3/protocol.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/protocol.proto:369]: nit: can you clarify what happens if the keep alive fails?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
API LGTM modulo small comments. I will take a look at the code once you sort out the CI issues.

/wait

**Comment by ggreenway** [comment]
This is related to #7136.

File: source/common/http/http2/codec_impl.cc
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc:572]: Does this timer need to be cancelled in any deferred delete close paths so we don't get any kind of race? That is a typical bug that we hit.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
This is cool. Flushing out a few impl comments from a first pass. Thank you!

/wait

File: source/common/http/http2/codec_impl.cc
Comment (ggreenway) [reply source/common/http/http2/codec_impl.cc:589]: It's useful for debugging. The h2 spec requires 8 bytes of payload in the PING. nghttp2 will fill it with zeros if it isn't supplied. I found it useful for matching up the PING and ACK in logs.

File: source/common/http/http2/codec_impl.cc
Comment (ggreenway) [reply source/common/http/http2/codec_impl.cc:603]: Yes, that's exactly the case I was worried about. I'll comment.

File: source/common/http/http2/codec_impl.cc
Comment (ggreenway) [reply source/common/http/http2/codec_impl.cc:612]: If it was outside the `if`, the timer could be nullptr and we crash.

File: source/common/http/http2/codec_impl.cc
Comment (ggreenway) [reply source/common/http/http2/codec_impl.cc:572]: I grep'd through all `deferredDelete()`s in `source/common/http`, and it looks like only streams are deferDeleted, not connections.

File: source/common/http/http2/codec_impl.cc
Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:590]: OK that's fine. Can you add comments about what this is for?

File: source/common/http/http2/codec_impl.cc
Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:614]: Sorry I read this too quickly while doing something else. This makes sense.

File: source/common/http/http2/codec_impl.cc
Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:572]: OK sounds good.

File: source/common/http/http2/codec_impl.cc
Comment (mattklein123) [reply source/common/http/http2/codec_impl.cc:742]: Can you add the comment here about that this is for debugging and should match the time stamp?

**Comment by mattklein123** [comment]
See also https://github.com/envoyproxy/envoy/pull/13245 in case that will make this change less verbose. cc @akonradi

File: api/envoy/config/core/v3/protocol.proto
Comment (htuch) [inline api/envoy/config/core/v3/protocol.proto:198]: I think this should be `type.v3.Percent`, similar to what we do in `api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto`

Comment (ggreenway) [reply api/envoy/config/core/v3/protocol.proto:198]: I didn't know that type existed. I'll switch to it.

**Comment by akonradi** [comment]
It looks like #13245 won't help here since `Api::Api` isn't already an existing argument to any of the methods that this adds `RandomGenerator` to.

**Review by mattklein123 — APPROVED** [review]
LGTM, thanks!

File: test/common/http/http2/codec_impl_test.cc
Comment (yanavlasov) [inline test/common/http/http2/codec_impl_test.cc:900]: The comment is not correct, since the PING frame is not delivered to the peer in this test.
To make PING frames flow you will need to mock the `write` method on the `server_connection_`, like so:
```C++
  ON_CALL(server_connection_, write(_, _))
      .WillByDefault(
          Invoke([&](Buffer::Instance& data, bool) -> void { 
            // Peek into the data to check if this is indeed a PING frame
            client_wrapper_.buffer_.add(data); 
          }));
```

You can then dispatch the PING frame to the client codec like so `client_wrapper_.dispatch(Buffer::OwnedImpl(), *client_);`
Client codec will produce PING ACK in response and you can send it back to the server with the `client_->sendPendingFrames()` method and check that the ACK timeout timer was disabled.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: test/common/http/http2/codec_impl_test.cc
Comment (ggreenway) [reply test/common/http/http2/codec_impl_test.cc:900]: When I started writing the test, I expected the behavior you described. But it is definitely sending a PING and receiving an ACK. Validated in the debugger. Which is why the expectations below are met.

File: test/common/http/http2/codec_impl_test.cc
Comment (ggreenway) [reply test/common/http/http2/codec_impl_test.cc:901]: It's also testing send/receive of keepalives. How about `ConnectionKeepalive`, which matches the proto setting used to configure it?

File: test/common/http/http2/codec_impl_test.cc
Comment (ggreenway) [reply test/common/http/http2/codec_impl_test.cc:900]: Traced cause back to line 160 of this file:

```
  void setupDefaultConnectionMocks() {
    ON_CALL(client_connection_, write(_, _))
        .WillByDefault(Invoke([&](Buffer::Instance& data, bool) -> void {
          if (corrupt_metadata_frame_) {
            corruptMetadataFramePayload(data);
          }
          auto status = server_wrapper_.dispatch(data, *server_);
          if (!status.ok()) {
            throw ServerCodecError(std::move(status));
          }
        }));
    ON_CALL(server_connection_, write(_, _))
        .WillByDefault(Invoke([&](Buffer::Instance& data, bool) -> void {
          auto status = client_wrapper_.dispatch(data, *client_);
          if (!status.ok()) {
            throw ClientCodecError(std::move(status));
          }
        }));
  }
```

**Comment by ggreenway** [comment]
Got another clang_tidy error about not being able to find a header on the last run. Hoping that merging master will make it pass. No idea why it failed, since no related files changed.

**Comment by mattklein123** [comment]
If you merge main it should fix gcc. I think your other issues are an AZP outage.

/wait

File: docs/root/version_history/current.rst
Comment (htuch) [inline docs/root/version_history/current.rst:92]: Re: how to hook this into documentation to folks know to use it. I suggest that we update all examples that use xDS in the tree to use keepalive. Also, we can provide guidance in https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api and https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/grpc_service.proto#envoy-v3-api-msg-config-core-v3-grpcservice-envoygrpc.

We could also add a warning potentially when Envoy gRPC is used for xDS without this option enabled.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: docs/root/version_history/current.rst
Comment (ggreenway) [reply docs/root/version_history/current.rst:92]: Added docs.

Comment (ggreenway) [reply docs/root/version_history/current.rst:92]: I didn't add any to `grpc_service` because that applies to things other than xds, so it didn't seem like the right place for this.

File: api/xds_protocol.rst
Comment (mattklein123) [inline api/xds_protocol.rst:703]: These seem pretty aggressive. I would go either a longer interval and longer timeout? Same elsewhere.

File: api/xds_protocol.rst
Comment (mattklein123) [inline api/xds_protocol.rst:703]: Do we want to also demonstrate TCP keep alive and describe both options? I think TCP keep alive would work fine in many cases and is lower overhead.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small comments.

/wait

File: docs/root/version_history/current.rst
Comment (htuch) [reply docs/root/version_history/current.rst:92]: I agree this is more general, but I think that when you do opt to use Envoy gRPC, you should set this as best practice, hence why I think it's good to provide a recommendation in those docs.

Comment (htuch) [reply docs/root/version_history/current.rst:92]: Actually, ignore my last comment, I see where you've added the best practices, that's all good.

File: api/xds_protocol.rst
Comment (ggreenway) [reply api/xds_protocol.rst:703]: It's just an example; people should come up with their own values. But if you want it changed, what values should we use?

File: api/xds_protocol.rst
Comment (ggreenway) [reply api/xds_protocol.rst:703]: Yeah, that makes sense. I'll update.

File: api/xds_protocol.rst
Comment (mattklein123) [reply api/xds_protocol.rst:703]: People just copy the docs... I would probably use 30s interval with 5s timeout?

Comment (ggreenway) [reply api/xds_protocol.rst:706]: updated

**Comment by ggreenway** [comment]
I don't understand why, but `//test/config_test:example_configs_test` starting failing after I touched the examples, because they use deprecated field `watchdog`. That field was deprecated a week ago; why did it start failing now? Was bazel caching the test result based only on the content of the example .yaml, and not re-running when Envoy changed? @htuch

**Comment by mattklein123** [comment]
> I don't understand why, but //test/config_test:example_configs_test starting failing after I touched the examples, because they use deprecated field watchdog. That field was deprecated a week ago; why did it start failing now? Was bazel caching the test result based only on the content of the example .yaml, and not re-running when Envoy changed? @htuch

Hmm interesting. cc @phlax who has been working on this.

**Comment by mattklein123** [comment]
LGTM. Can you merge main one more time which should fix ASAN? I don't understand the example config issue but we can look at that separately. I think it must have to do with the fact that watchdog is a bootstrap field, but I don't understand why the failure would not be deterministic.

/wait

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
